File: packages/react-native/ios/ReportDirectory.swift
// capture-es - bitdrift's ES SDK
// Copyright Bitdrift, Inc. All rights reserved.
//
// Use of this source code is governed by a source available license that can be found in the
// LICENSE file or at:
// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt

import Foundation

enum ReportDirectory {
    private static let fileManager = FileManager.default
    private static let watcherDirPath = "reports/watcher"
    
    static func sdkBaseDirectory() -> URL? {
        return try? fileManager.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
    }
    
    static func ensureSDKDirectory() -> URL? {
        guard let baseDir = sdkBaseDirectory() else {
            return nil
        }
        
        let sdkDir = baseDir.appendingPathComponent("bitdrift_capture", isDirectory: true)
        
        if !fileManager.fileExists(atPath: sdkDir.path) {
            try? fileManager.createDirectory(at: sdkDir, withIntermediateDirectories: true)
        }
        
        return sdkDir
    }
    
    static func getWatcherDirectory() -> URL? {
        return ensureSDKDirectory()?.appendingPathComponent(watcherDirPath, isDirectory: true)
    }
    
    static func setupWatcherDirectory(enableJsErrors: Bool) {
        guard let sdkDir = ensureSDKDirectory() else {
            return
        }
        let watcherDir = sdkDir.appendingPathComponent(watcherDirPath, isDirectory: true)
        
        if enableJsErrors {
            if !fileManager.fileExists(atPath: watcherDir.path) {
                try? fileManager.createDirectory(at: watcherDir, withIntermediateDirectories: true)
            }
        } else {
            if fileManager.fileExists(atPath: watcherDir.path) {
                try? fileManager.removeItem(at: watcherDir)
            }
        }
    }
    
    static func reportsDirectory(isFatal: Bool) -> URL? {
        guard let sdkDir = ensureSDKDirectory() else {
            return nil
        }
        
        let reportsDir = sdkDir.appendingPathComponent(
            isFatal ? "reports/new" : "reports/watcher/current_session",
            isDirectory: true
        )
        
        if !fileManager.fileExists(atPath: reportsDir.path) {
            try? fileManager.createDirectory(at: reportsDir, withIntermediateDirectories: true)
        }
        
        guard fileManager.isWritableFile(atPath: reportsDir.path) else {
            return nil
        }
        
        return reportsDir
    }
    
    static func destinationPath(isFatal: Bool) -> String? {
        guard let reportsDir = reportsDirectory(isFatal: isFatal) else {
            return nil
        }
        
        let fileName = isFatal
            ? "javascript-fatal-\(UUID().uuidString).cap"
            : "javascript-non-fatal-\(UUID().uuidString).cap"
        
        return reportsDir.appendingPathComponent(fileName).path
    }
}

File: packages/react-native/android/src/main/java/com/bdreactnative/ReportDirectory.kt
// capture-es - bitdrift's ES SDK
// Copyright Bitdrift, Inc. All rights reserved.
//
// Use of this source code is governed by a source available license that can be found in the
// LICENSE file or at:
// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt

package com.bdreactnative

import android.content.Context
import android.util.Log
import io.bitdrift.capture.utils.SdkDirectory
import java.io.File

object ReportDirectory {
    private const val TAG = "BdReactNative"
    private const val WATCHER_DIR_PATH = "reports/watcher"

    /**
     * Returns the watcher directory path for JavaScript error reports
     */
    fun getWatcherDirectory(context: Context): File {
        return File(SdkDirectory.getPath(context), WATCHER_DIR_PATH)
    }

    /**
     * Sets up the report watcher directory based on whether JS errors are enabled
     * Creates the directory if enabled, removes it if disabled
     */
    fun setupWatcherDirectory(context: Context, enableJsErrors: Boolean) {
        runCatching {
            val watcherDir = getWatcherDirectory(context)

            if (enableJsErrors) {
                createWatcherDirectory(watcherDir)
            } else {
                removeWatcherDirectory(watcherDir)
            }
        }.onFailure { error ->
            Log.w(TAG, "Failed to handle directory modifications for enableJsErrors $enableJsErrors: $error")
        }
    }

    /**
     * Creates the watcher directory if it doesn't exist
     */
    private fun createWatcherDirectory(watcherDir: File) {
        if (!watcherDir.exists() && !watcherDir.mkdirs()) {
            Log.w(TAG, "Failed to create reports/watcher directory")
        }
    }

    /**
     * Removes the watcher directory if it exists
     */
    private fun removeWatcherDirectory(watcherDir: File) {
        if (watcherDir.exists() && !watcherDir.deleteRecursively()) {
            Log.w(TAG, "Failed to delete reports/watcher directory")
        }
    }
}

